File: ruby/bob/bob.rb
class Bob
  PHRASES = {
    yell:      'Whoa, chill out!',
    question:  'Sure.',
    silence:   'Fine. Be that way!',
    whatever:  'Whatever.'
  }

  def hey(phrase)
    phrase = phrase.gsub("\n", " ")
    return PHRASES[:silence]  if is_silence(phrase)
    return PHRASES[:yell]     if is_yelling(phrase)
    return PHRASES[:question] if is_question(phrase)
    return PHRASES[:whatever]
  end

  def is_yelling(phrase)
    !(phrase =~ /[a-zA-Z]/).nil? && phrase == phrase.upcase
  end

  def is_question(phrase)
    !(phrase =~ /\?$/).nil?
  end

  def is_silence(phrase)
    (phrase =~ /[a-zA-Z0-9]/).nil?
  end
end

File: elixir/bob/bob.exs
defmodule Bob do

  @phrases %{
    yell:      "Whoa, chill out!",
    question:  "Sure.",
    silence:   "Fine. Be that way!",
    whatever:  "Whatever."
  }

  def hey(input) do
    res = @phrases[:whatever]
    if is_question?(input), do: res = @phrases[:question]
    if is_yelling?(input),  do: res = @phrases[:yell]
    if is_silence?(input),  do: res = @phrases[:silence]
    res
  end

  def is_yelling?(input) do
    input == String.upcase(input) && not is_question?(input) && not only_nums_and_chars(input)
  end

  def only_nums_and_chars(input) do
    not Regex.match?(~r/[[:alpha:]]/u, input)
  end

  def is_question?(input) do
    Regex.match?(~r/\?$/, input)
  end

  def is_silence?(input) do
    not Regex.match?(~r/[[:alnum:]]/u, input)
  end

end
